File: compiler/packages/babel-plugin-react-forget/src/ReactiveScopes/MergeReactiveScopesThatInvalidateTogether.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { CompilerError } from "..";
import {
  IdentifierId,
  InstructionId,
  Place,
  ReactiveBlock,
  ReactiveFunction,
  ReactiveInstruction,
  ReactiveScope,
  ReactiveScopeBlock,
  ReactiveScopeDependencies,
  ReactiveScopeDependency,
  ReactiveStatement,
  makeInstructionId,
} from "../HIR";
import { assertExhaustive } from "../Utils/utils";
import { printReactiveScopeSummary } from "./PrintReactiveFunction";
import {
  ReactiveFunctionTransform,
  ReactiveFunctionVisitor,
  Transformed,
  visitReactiveFunction,
} from "./visitors";

/**
 * The primary goal of this pass is to reduce memoization overhead, specifically:
 * - Use fewer memo slots
 * - Reduce the number of comparisons and other memoization-related instructions
 *
 * The algorithm merges in two main cases: consecutive scopes that invalidate together
 * or nested scopes that invalidate together
 *
 * ## Consecutive Scopes
 *
 * The idea is that if two consecutive scopes would always invalidate together,
 * it's more efficient to group the scopes together to save on memoization overhead.
 *
 * This optimization is necessarily somewhat limited. First, we only merge
 * scopes that are in the same (reactive) block, ie we don't merge across
 * control-flow or block-scoping boundaries. Second, we can only merge scopes
 * so long as any intermediate instructions are safe to memoize — specifically,
 * as long as the values created by those instructions are only referenced by
 * the second scope and not elsewhere. This is to avoid changing control-flow
 * and to avoid increasing the number of scope outputs (which defeats the optimization).
 *
 * With that in mind we can apply the optimization in two cases. Given a block with
 * scope A, some safe-to-memoize instructions I, and scope B, we can merge scopes when:
 * - A and B have identical dependencies. This means they will invalidate together, so
 *   by merging the scopes we can avoid duplicate cache slots and duplicate checks of
 *   those dependencies.
 * - The output of A is the input to B. Any invalidation of A will change its output
 *   which invalidates B, so we can similarly merge scopes. Note that this optimization
 *   may not be beneficial if the outupts of A are not guaranteed to change if its input
 *   changes, but in practice this is generally the case.
 *
 * ## Nested Scopes
 *
 * In this case, if an inner scope has the same dependencies as its parent, then we can
 * flatten away the inner scope since it will always invalidate at the same time.
 *
 * Note that PropagateScopeDependencies propagates scope dependencies upwards. This ensures
 * that parent scopes have the union of their own direct dependencies as well as those of
 * their (transitive) children. As a result nested scopes may have the same or fewer
 * dependencies than their parents, but not more dependencies. If they have fewer dependncies,
 * it means that the inner scope does not always invalidate with the parent and we should not
 * flatten. If they inner scope has the exact same dependencies, however, then it's always
 * better to flatten.
 */
export function mergeReactiveScopesThatInvalidateTogether(
  fn: ReactiveFunction
): void {
  const lastUsageVisitor = new FindLastUsageVisitor();
  visitReactiveFunction(fn, lastUsageVisitor, undefined);
  visitReactiveFunction(fn, new Transform(lastUsageVisitor.lastUsage), null);
}

const DEBUG: boolean = false;
function log(msg: string): void {
  if (DEBUG) {
    console.log(msg);
  }
}

class FindLastUsageVisitor extends ReactiveFunctionVisitor<void> {
  lastUsage: Map<IdentifierId, InstructionId> = new Map();

  override visitPlace(id: InstructionId, place: Place, _state: void): void {
    const previousUsage = this.lastUsage.get(place.identifier.id);
    const lastUsage =
      previousUsage !== undefined
        ? makeInstructionId(Math.max(previousUsage, id))
        : id;
    this.lastUsage.set(place.identifier.id, lastUsage);
  }
}

class Transform extends ReactiveFunctionTransform<ReactiveScopeDependencies | null> {
  lastUsage: Map<IdentifierId, InstructionId>;

  constructor(lastUsage: Map<IdentifierId, InstructionId>) {
    super();
    this.lastUsage = lastUsage;
  }

  override transformScope(
    scope: ReactiveScopeBlock,
    state: ReactiveScopeDependencies | null
  ): Transformed<ReactiveStatement> {
    this.visitScope(scope, scope.scope.dependencies);
    if (
      state !== null &&
      areEqualDependencies(state, scope.scope.dependencies)
    ) {
      return { kind: "replace-many", value: scope.instructions };
    } else {
      return { kind: "keep" };
    }
  }

  override visitBlock(
    block: ReactiveBlock,
    state: ReactiveScopeDependencies | null
  ): void {
    // Pass 1: visit nested blocks to potentially merge their scopes
    this.traverseBlock(block, state);

    // Pass 2: identify scopes for merging
    type MergedScope = {
      scope: ReactiveScopeBlock;
      from: number;
      to: number;
      lvalues: Set<IdentifierId>;
    };
    let current: MergedScope | null = null;
    const merged: Array<MergedScope> = [];
    function reset(): void {
      CompilerError.invariant(current !== null, {
        loc: null,
        reason:
          "MergeConsecutiveScopes: expected current scope to be non-null if reset()",
        suggestions: null,
        description: null,
      });
      if (current.to > current.from + 1) {
        merged.push(current);
      }
      current = null;
    }
    for (let i = 0; i < block.length; i++) {
      const instr = block[i]!;
      switch (instr.kind) {
        case "terminal": {
          // For now we don't merge across terminals
          if (current !== null) {
            log(
              `Reset scope @${current.scope.scope.id} from terminal [${instr.terminal.id}]`
            );
            reset();
          }
          break;
        }
        case "instruction": {
          switch (instr.instruction.value.kind) {
            case "ComputedLoad":
            case "JSXText":
            case "LoadLocal":
            case "Primitive":
            case "PropertyLoad": {
              // We can merge two scopes if there are intervening instructions, but:
              // - Only if the instructions are simple and it's okay to make them
              //   execute conditionally (hence allowing a conservative subset of value kinds)
              // - The values produced are used at or before the next scope. If they are used
              //   later and we move them into the scope, then they wouldn't be accessible to
              //   subsequent code wo expanding the set of declarations, which we want to avoid
              if (current !== null && instr.instruction.lvalue !== null) {
                current.lvalues.add(instr.instruction.lvalue.identifier.id);
              }
              break;
            }
            default: {
              // Other instructions are known to prevent merging, so we reset the scope if present
              if (current !== null) {
                log(
                  `Reset scope @${current.scope.scope.id} from instruction [${instr.instruction.id}]`
                );
                reset();
              }
            }
          }
          break;
        }
        case "scope": {
          if (
            current !== null &&
            canMergeScopes(current.scope, instr) &&
            areLValuesLastUsedByScope(
              instr.scope,
              current.lvalues,
              this.lastUsage
            )
          ) {
            // The current and next scopes can merge!
            log(
              `Can merge scope @${current.scope.scope.id} with @${instr.scope.id}`
            );
            // Update the merged scope's range
            current.scope.scope.range.end = makeInstructionId(
              Math.max(current.scope.scope.range.end, instr.scope.range.end)
            );
            // Add declarations
            for (const [key, value] of instr.scope.declarations) {
              current.scope.scope.declarations.set(key, value);
            }
            // Then prune declarations - this removes declarations from the earlier
            // scope that are last-used at or before the newly merged subsequent scope
            updateScopeDeclarations(current.scope.scope, this.lastUsage);
            current.to = i + 1;
            // We already checked that intermediate values were used at-or-before the merged
            // scoped, so we can reset
            current.lvalues.clear();

            if (!scopeAlwaysInvalidatesOnDependencyChanges(instr)) {
              // The subsequent scope that we just merged isn't guaranteed to invalidate if its
              // inputs change, so it is not a candidate for future merging
              log(
                `  but scope @${instr.scope.id} doesnt guaranteed invalidate so it cannot merge further`
              );
              reset();
            }
          } else {
            // No previous scope, or the scope cannot merge
            if (current !== null) {
              // Reset if necessary
              log(
                `Reset scope @${current.scope.scope.id}, not mergeable with subsequent scope @${instr.scope.id}`
              );
              reset();
            }
            // Only set a new merge candidate if the scope is guaranteed to invalidate on changes
            if (scopeAlwaysInvalidatesOnDependencyChanges(instr)) {
              current = {
                scope: instr,
                from: i,
                to: i + 1,
                lvalues: new Set(),
              };
            } else {
              log(
                `scope @${instr.scope.id} doesnt guaranteed invalidate so it cannot merge further`
              );
            }
          }
          break;
        }
        default: {
          assertExhaustive(
            instr,
            `Unexpected instruction kind '${(instr as any).kind}'`
          );
        }
      }
    }
    if (current !== null) {
      reset();
    }
    if (merged.length) {
      log(`merged ${merged.length} scopes:`);
      for (const entry of merged) {
        log(
          printReactiveScopeSummary(entry.scope.scope) +
            ` from=${entry.from} to=${entry.to}`
        );
      }
    }

    // Pass 3: optional: if scopes can be merged, merge them and update the block
    if (merged.length === 0) {
      // Nothing merged, nothing to do!
      return;
    }
    const nextInstructions = [];
    let index = 0;
    for (const entry of merged) {
      if (index < entry.from) {
        nextInstructions.push(...block.slice(index, entry.from));
        index = entry.from;
      }
      const mergedScope = block[entry.from]!;
      CompilerError.invariant(mergedScope.kind === "scope", {
        loc: null,
        reason:
          "MergeConsecutiveScopes: Expected scope starting index to be a scope",
        description: null,
        suggestions: null,
      });
      nextInstructions.push(mergedScope);
      index++;
      while (index < entry.to) {
        const instr = block[index++]!;
        if (instr.kind === "scope") {
          mergedScope.instructions.push(...instr.instructions);
        } else {
          mergedScope.instructions.push(instr);
        }
      }
    }
    while (index < block.length) {
      nextInstructions.push(block[index++]!);
    }
    block.length = 0;
    block.push(...nextInstructions);
  }
}

/**
 * Updates @param scope's declarations to remove any declarations that are not
 * used after the scope, based on the scope's updated range post-merging.
 */
function updateScopeDeclarations(
  scope: ReactiveScope,
  lastUsage: Map<IdentifierId, InstructionId>
): void {
  for (const [key] of scope.declarations) {
    const lastUsedAt = lastUsage.get(key)!;
    if (lastUsedAt < scope.range.end) {
      scope.declarations.delete(key);
    }
  }
}

/**
 * Returns whether the given @param scope is the last usage of all
 * the given @param lvalues. Returns false if any of the lvalues
 * are used again after the scope.
 */
function areLValuesLastUsedByScope(
  scope: ReactiveScope,
  lvalues: Set<IdentifierId>,
  lastUsage: Map<IdentifierId, InstructionId>
): boolean {
  for (const lvalue of lvalues) {
    const lastUsedAt = lastUsage.get(lvalue)!;
    if (lastUsedAt >= scope.range.end) {
      log(`  lvalue ${lvalue} used after scope @${scope.id}, cannot merge`);
      return false;
    }
  }
  return true;
}

function canMergeScopes(a: ReactiveScopeBlock, b: ReactiveScopeBlock): boolean {
  // Don't merge scopes with reassignments
  if (a.scope.reassignments.size !== 0 || b.scope.reassignments.size !== 0) {
    log(`  cannot merge, has reassignments`);
    return false;
  }
  // Merge scopes whose dependencies are identical
  if (areEqualDependencies(a.scope.dependencies, b.scope.dependencies)) {
    log(`  canMergeScopes: dependencies are equal`);
    return true;
  }
  // Merge scopes where the outputs of the previous scope are the inputs
  // of the subsequent scope. Note that the output of a scope is not
  // guaranteed to change when its inputs change, for example `foo(x)`
  // may not change when `x` changes, for example `foo(x) { return x < 10}`
  // will not change as x changes from 0 -> 1.
  // Therefore we check that the outputs of the previous scope are of a type
  // that is guaranteed to invalidate with its inputs, and only merge in this case.
  if (
    areEqualDependencies(
      new Set(
        [...a.scope.declarations.values()].map((declaration) => ({
          identifier: declaration.identifier,
          path: [],
        }))
      ),
      b.scope.dependencies
    )
  ) {
    log(`  outputs of prev are input to current`);
    return true;
  }
  log(`  cannot merge scopes:`);
  log(`  ${printReactiveScopeSummary(a.scope)}`);
  log(`  ${printReactiveScopeSummary(b.scope)}`);
  return false;
}

function areEqualDependencies(
  a: Set<ReactiveScopeDependency>,
  b: Set<ReactiveScopeDependency>
): boolean {
  if (a.size !== b.size) {
    return false;
  }
  for (const aValue of a) {
    let found = false;
    for (const bValue of b) {
      if (
        aValue.identifier === bValue.identifier &&
        areEqualPaths(aValue.path, bValue.path)
      ) {
        found = true;
        break;
      }
    }
    if (!found) {
      return false;
    }
  }
  return true;
}

function areEqualPaths(a: Array<string>, b: Array<string>): boolean {
  return a.length === b.length && a.every((item, ix) => item === b[ix]);
}

function scopeAlwaysInvalidatesOnDependencyChanges(
  scope: ReactiveScopeBlock
): boolean {
  const visitor = new DeclarationTypeVisitor(scope.scope);
  visitor.visitScope(scope, undefined);
  return visitor.alwaysInvalidatesOnInputChange;
}

class DeclarationTypeVisitor extends ReactiveFunctionVisitor<void> {
  scope: ReactiveScope;
  alwaysInvalidatesOnInputChange: boolean = false;

  constructor(scope: ReactiveScope) {
    super();
    this.scope = scope;
  }

  override visitScope(scope: ReactiveScopeBlock, state: void): void {
    if (scope.scope.id !== this.scope.id) {
      return;
    }
    this.traverseScope(scope, state);
  }

  override visitInstruction(
    instruction: ReactiveInstruction,
    state: void
  ): void {
    this.traverseInstruction(instruction, state);
    if (
      instruction.lvalue === null ||
      !this.scope.declarations.has(instruction.lvalue.identifier.id)
    ) {
      // no lvalue or this instruction isn't directly constructing a
      // scope output value, skip
      log(
        `    skip instruction lvalue=${
          instruction.lvalue?.identifier.id
        } declaration?=${
          instruction.lvalue != null &&
          this.scope.declarations.has(instruction.lvalue.identifier.id)
        } scope=${printReactiveScopeSummary(this.scope)}`
      );
      return;
    }
    switch (instruction.value.kind) {
      case "FunctionExpression":
      case "ArrayExpression":
      case "JsxExpression":
      case "JsxFragment":
      case "ObjectExpression": {
        // These instruction types *always* allocate. If they execute
        // they will produce a new value, triggering downstream reactive
        // updates
        this.alwaysInvalidatesOnInputChange = true;
        break;
      }
    }
  }
}

File: compiler/crates/react_compiler_reactive_scopes/src/merge_reactive_scopes_that_invalidate_together.rs
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

//! MergeReactiveScopesThatInvalidateTogether — merges adjacent or nested scopes
//! that share dependencies (and thus invalidate together) to reduce memoization overhead.
//!
//! Corresponds to `src/ReactiveScopes/MergeReactiveScopesThatInvalidateTogether.ts`.

use std::collections::{HashMap, HashSet};

use react_compiler_diagnostics::CompilerError;
use react_compiler_hir::{
    DeclarationId, DependencyPathEntry, EvaluationOrder, InstructionKind, InstructionValue, Place,
    ReactiveBlock, ReactiveFunction, ReactiveStatement, ReactiveValue,
    ReactiveScopeBlock, ReactiveScopeDependency, ScopeId, Type,
    environment::Environment,
    object_shape::{BUILT_IN_ARRAY_ID, BUILT_IN_FUNCTION_ID, BUILT_IN_JSX_ID, BUILT_IN_OBJECT_ID},
};

use crate::visitors::{
    ReactiveFunctionTransform, ReactiveFunctionVisitor, Transformed, transform_reactive_function,
    visit_reactive_function,
};

// =============================================================================
// Public entry point
// =============================================================================

/// Merges adjacent reactive scopes that share dependencies (invalidate together).
/// TS: `mergeReactiveScopesThatInvalidateTogether`
pub fn merge_reactive_scopes_that_invalidate_together(
    func: &mut ReactiveFunction,
    env: &mut Environment,
) -> Result<(), CompilerError> {
    // Pass 1: find last usage of each declaration
    let visitor = FindLastUsageVisitor { env: &*env };
    let mut last_usage: HashMap<DeclarationId, EvaluationOrder> = HashMap::new();
    visit_reactive_function(func, &visitor, &mut last_usage);

    // Pass 2+3: merge scopes
    let mut transform = MergeTransform {
        env,
        last_usage,
        temporaries: HashMap::new(),
    };
    let mut state: Option<Vec<ReactiveScopeDependency>> = None;
    transform_reactive_function(func, &mut transform, &mut state)
}

// =============================================================================
// Pass 1: FindLastUsageVisitor
// =============================================================================

/// TS: `class FindLastUsageVisitor extends ReactiveFunctionVisitor<void>`
struct FindLastUsageVisitor<'a> {
    env: &'a Environment,
}

impl<'a> ReactiveFunctionVisitor for FindLastUsageVisitor<'a> {
    type State = HashMap<DeclarationId, EvaluationOrder>;

    fn env(&self) -> &Environment {
        self.env
    }

    fn visit_place(&self, id: EvaluationOrder, place: &Place, state: &mut Self::State) {
        let decl_id = self.env.identifiers[place.identifier.0 as usize].declaration_id;
        let entry = state.entry(decl_id).or_insert(id);
        if id > *entry {
            *entry = id;
        }
    }
}

// =============================================================================
// Pass 2+3: MergeTransform
// =============================================================================

/// TS: `class Transform extends ReactiveFunctionTransform<ReactiveScopeDependencies | null>`
struct MergeTransform<'a> {
    env: &'a mut Environment,
    last_usage: HashMap<DeclarationId, EvaluationOrder>,
    temporaries: HashMap<DeclarationId, DeclarationId>,
}

impl<'a> ReactiveFunctionTransform for MergeTransform<'a> {
    type State = Option<Vec<ReactiveScopeDependency>>;

    fn env(&self) -> &Environment {
        self.env
    }

    /// TS: `override transformScope(scopeBlock, state)`
    fn transform_scope(
        &mut self,
        scope: &mut ReactiveScopeBlock,
        state: &mut Self::State,
    ) -> Result<Transformed<ReactiveStatement>, CompilerError> {
        let scope_deps = self.env.scopes[scope.scope.0 as usize].dependencies.clone();
        // Save parent state and recurse with this scope's deps as state
        let parent_state = state.take();
        *state = Some(scope_deps.clone());
        self.visit_scope(scope, state)?;
        // Restore parent state
        *state = parent_state;

        // If parent has deps and they match, flatten the inner scope
        if let Some(parent_deps) = state.as_ref() {
            if are_equal_dependencies(parent_deps, &scope_deps, self.env) {
                let instructions = std::mem::take(&mut scope.instructions);
                return Ok(Transformed::ReplaceMany(instructions));
            }
        }
        Ok(Transformed::Keep)
    }

    /// TS: `override visitBlock(block, state)`
    fn visit_block(
        &mut self,
        block: &mut ReactiveBlock,
        state: &mut Self::State,
    ) -> Result<(), CompilerError> {
        // Pass 1: traverse nested (scope flattening handled by transform_scope)
        self.traverse_block(block, state)?;
        // Pass 2+3: merge consecutive scopes in this block
        self.merge_scopes_in_block(block)?;
        Ok(())
    }
}

impl<'a> MergeTransform<'a> {
    /// Identify and merge consecutive scopes that invalidate together.
    fn merge_scopes_in_block(&mut self, block: &mut ReactiveBlock) -> Result<(), CompilerError> {
        // Pass 2: identify scopes for merging
        struct MergedScope {
            scope_id: ScopeId,
            from: usize,
            to: usize,
            lvalues: HashSet<DeclarationId>,
        }

        let mut current: Option<MergedScope> = None;
        let mut merged: Vec<MergedScope> = Vec::new();

        let block_len = block.len();
        for i in 0..block_len {
            match &block[i] {
                ReactiveStatement::Terminal(_) => {
                    // Don't merge across terminals
                    if let Some(c) = current.take() {
                        if c.to > c.from + 1 {
                            merged.push(c);
                        }
                    }
                }
                ReactiveStatement::PrunedScope(_) => {
                    // Don't merge across pruned scopes
                    if let Some(c) = current.take() {
                        if c.to > c.from + 1 {
                            merged.push(c);
                        }
                    }
                }
                ReactiveStatement::Instruction(instr) => {
                    match &instr.value {
                        ReactiveValue::Instruction(iv) => {
                            match iv {
                                InstructionValue::BinaryExpression { .. }
                                | InstructionValue::ComputedLoad { .. }
                                | InstructionValue::JSXText { .. }
                                | InstructionValue::LoadGlobal { .. }
                                | InstructionValue::LoadLocal { .. }
                                | InstructionValue::Primitive { .. }
                                | InstructionValue::PropertyLoad { .. }
                                | InstructionValue::TemplateLiteral { .. }
                                | InstructionValue::UnaryExpression { .. } => {
                                    if let Some(ref mut c) = current {
                                        if let Some(lvalue) = &instr.lvalue {
                                            let decl_id = self.env.identifiers
                                                [lvalue.identifier.0 as usize]
                                                .declaration_id;
                                            c.lvalues.insert(decl_id);
                                            if let InstructionValue::LoadLocal { place, .. } = iv {
                                                let src_decl = self.env.identifiers
                                                    [place.identifier.0 as usize]
                                                    .declaration_id;
                                                self.temporaries.insert(decl_id, src_decl);
                                            }
                                        }
                                    }
                                }
                                InstructionValue::StoreLocal { lvalue, value, .. } => {
                                    if let Some(ref mut c) = current {
                                        if lvalue.kind == InstructionKind::Const {
                                            // Add the instruction lvalue (if any)
                                            if let Some(instr_lvalue) = &instr.lvalue {
                                                let decl_id = self.env.identifiers
                                                    [instr_lvalue.identifier.0 as usize]
                                                    .declaration_id;
                                                c.lvalues.insert(decl_id);
                                            }
                                            // Add the StoreLocal's lvalue place
                                            let store_decl = self.env.identifiers
                                                [lvalue.place.identifier.0 as usize]
                                                .declaration_id;
                                            c.lvalues.insert(store_decl);
                                            // Track temporary mapping
                                            let value_decl = self.env.identifiers
                                                [value.identifier.0 as usize]
                                                .declaration_id;
                                            let mapped = self
                                                .temporaries
                                                .get(&value_decl)
                                                .copied()
                                                .unwrap_or(value_decl);
                                            self.temporaries.insert(store_decl, mapped);
                                        } else {
                                            // Non-const StoreLocal — reset
                                            let c = current.take().unwrap();
                                            if c.to > c.from + 1 {
                                                merged.push(c);
                                            }
                                        }
                                    }
                                }
                                _ => {
                                    // Other instructions prevent merging
                                    if let Some(c) = current.take() {
                                        if c.to > c.from + 1 {
                                            merged.push(c);
                                        }
                                    }
                                }
                            }
                        }
                        _ => {
                            // Non-Instruction reactive values prevent merging
                            if let Some(c) = current.take() {
                                if c.to > c.from + 1 {
                                    merged.push(c);
                                }
                            }
                        }
                    }
                }
                ReactiveStatement::Scope(scope_block) => {
                    let next_scope_id = scope_block.scope;
                    if let Some(ref mut c) = current {
                        let current_scope_id = c.scope_id;
                        if can_merge_scopes(
                            current_scope_id,
                            next_scope_id,
                            self.env,
                            &self.temporaries,
                        ) && are_lvalues_last_used_by_scope(
                            next_scope_id,
                            &c.lvalues,
                            &self.last_usage,
                            self.env,
                        ) {
                            // Merge: extend the current scope's range
                            let next_range_end =
                                self.env.scopes[next_scope_id.0 as usize].range.end;
                            let current_range_end =
                                self.env.scopes[current_scope_id.0 as usize].range.end;
                            self.env.scopes[current_scope_id.0 as usize].range.end =
                                EvaluationOrder(current_range_end.0.max(next_range_end.0));

                            // Merge declarations from next into current
                            let next_decls =
                                self.env.scopes[next_scope_id.0 as usize].declarations.clone();
                            for (key, value) in next_decls {
                                let current_decls = &mut self.env.scopes
                                    [current_scope_id.0 as usize]
                                    .declarations;
                                if let Some(existing) =
                                    current_decls.iter_mut().find(|(k, _)| *k == key)
                                {
                                    existing.1 = value;
                                } else {
                                    current_decls.push((key, value));
                                }
                            }

                            // Prune declarations that are no longer used after the merged scope
                            update_scope_declarations(
                                current_scope_id,
                                &self.last_usage,
                                self.env,
                            );

                            c.to = i + 1;
                            c.lvalues.clear();

                            if !scope_is_eligible_for_merging(next_scope_id, self.env) {
                                let c = current.take().unwrap();
                                if c.to > c.from + 1 {
                                    merged.push(c);
                                }
                            }
                        } else {
                            // Cannot merge — reset
                            let c = current.take().unwrap();
                            if c.to > c.from + 1 {
                                merged.push(c);
                            }
                            // Start new candidate if eligible
                            if scope_is_eligible_for_merging(next_scope_id, self.env) {
                                current = Some(MergedScope {
                                    scope_id: next_scope_id,
                                    from: i,
                                    to: i + 1,
                                    lvalues: HashSet::new(),
                                });
                            }
                        }
                    } else {
                        // No current — start new candidate if eligible
                        if scope_is_eligible_for_merging(next_scope_id, self.env) {
                            current = Some(MergedScope {
                                scope_id: next_scope_id,
                                from: i,
                                to: i + 1,
                                lvalues: HashSet::new(),
                            });
                        }
                    }
                }
            }
        }
        // Flush remaining
        if let Some(c) = current.take() {
            if c.to > c.from + 1 {
                merged.push(c);
            }
        }

        // Pass 3: apply merges
        if merged.is_empty() {
            return Ok(());
        }

        let mut next_instructions: Vec<ReactiveStatement> = Vec::new();
        let mut index = 0;
        let all_stmts: Vec<ReactiveStatement> = std::mem::take(block);

        for entry in &merged {
            // Push everything before the merge range
            while index < entry.from {
                next_instructions.push(all_stmts[index].clone());
                index += 1;
            }
            // The first item in the merge range must be a scope
            let mut merged_scope = match &all_stmts[entry.from] {
                ReactiveStatement::Scope(s) => s.clone(),
                _ => {
                    return Err(react_compiler_diagnostics::CompilerDiagnostic::new(
                        react_compiler_diagnostics::ErrorCategory::Invariant,
                        "MergeConsecutiveScopes: Expected scope at starting index",
                        None,
                    )
                    .into());
                }
            };
            index += 1;
            while index < entry.to {
                let stmt = &all_stmts[index];
                index += 1;
                match stmt {
                    ReactiveStatement::Scope(inner_scope) => {
                        merged_scope
                            .instructions
                            .extend(inner_scope.instructions.clone());
                        self.env.scopes[merged_scope.scope.0 as usize]
                            .merged
                            .push(inner_scope.scope);
                    }
                    _ => {
                        merged_scope.instructions.push(stmt.clone());
                    }
                }
            }
            next_instructions.push(ReactiveStatement::Scope(merged_scope));
        }
        // Push remaining
        while index < all_stmts.len() {
            next_instructions.push(all_stmts[index].clone());
            index += 1;
        }

        *block = next_instructions;
        Ok(())
    }
}

// =============================================================================
// Helper functions
// =============================================================================

/// Updates scope declarations to remove any that are not used after the scope.
fn update_scope_declarations(
    scope_id: ScopeId,
    last_usage: &HashMap<DeclarationId, EvaluationOrder>,
    env: &mut Environment,
) {
    let range_end = env.scopes[scope_id.0 as usize].range.end;
    env.scopes[scope_id.0 as usize]
        .declarations
        .retain(|(_id, decl)| {
            let decl_declaration_id = env.identifiers[decl.identifier.0 as usize].declaration_id;
            match last_usage.get(&decl_declaration_id) {
                Some(last_used_at) => *last_used_at >= range_end,
                // If not tracked, keep the declaration (conservative)
                None => true,
            }
        });
}

/// Returns whether all lvalues are last used at or before the given scope.
fn are_lvalues_last_used_by_scope(
    scope_id: ScopeId,
    lvalues: &HashSet<DeclarationId>,
    last_usage: &HashMap<DeclarationId, EvaluationOrder>,
    env: &Environment,
) -> bool {
    let range_end = env.scopes[scope_id.0 as usize].range.end;
    for lvalue in lvalues {
        if let Some(&last_used_at) = last_usage.get(lvalue) {
            if last_used_at >= range_end {
                return false;
            }
        }
    }
    true
}

/// Check if two scopes can be merged.
fn can_merge_scopes(
    current_id: ScopeId,
    next_id: ScopeId,
    env: &Environment,
    temporaries: &HashMap<DeclarationId, DeclarationId>,
) -> bool {
    let current = &env.scopes[current_id.0 as usize];
    let next = &env.scopes[next_id.0 as usize];

    // Don't merge scopes with reassignments
    if !current.reassignments.is_empty() || !next.reassignments.is_empty() {
        return false;
    }

    // Merge scopes whose dependencies are identical
    if are_equal_dependencies(&current.dependencies, &next.dependencies, env) {
        return true;
    }

    // Merge scopes where outputs of current are inputs of next
    // Build synthetic dependencies from current's declarations
    let current_decl_deps: Vec<ReactiveScopeDependency> = current
        .declarations
        .iter()
        .map(|(_key, decl)| ReactiveScopeDependency {
            identifier: decl.identifier,
            reactive: true,
            path: Vec::new(),
            loc: None,
        })
        .collect();

    if are_equal_dependencies(&current_decl_deps, &next.dependencies, env) {
        return true;
    }

    // Check if all next deps have empty paths, always-invalidating types,
    // and correspond to current declarations (possibly through temporaries)
    if !next.dependencies.is_empty()
        && next.dependencies.iter().all(|dep| {
            if !dep.path.is_empty() {
                return false;
            }
            let dep_type =
                &env.types[env.identifiers[dep.identifier.0 as usize].type_.0 as usize];
            if !is_always_invalidating_type(dep_type) {
                return false;
            }
            let dep_decl = env.identifiers[dep.identifier.0 as usize].declaration_id;
            current.declarations.iter().any(|(_key, decl)| {
                let decl_decl_id = env.identifiers[decl.identifier.0 as usize].declaration_id;
                decl_decl_id == dep_decl
                    || temporaries.get(&dep_decl).copied() == Some(decl_decl_id)
            })
        })
    {
        return true;
    }

    false
}

/// Check if a type is always invalidating (guaranteed to change when inputs change).
pub fn is_always_invalidating_type(ty: &Type) -> bool {
    match ty {
        Type::Object { shape_id } => {
            if let Some(id) = shape_id {
                matches!(
                    id.as_str(),
                    s if s == BUILT_IN_ARRAY_ID
                        || s == BUILT_IN_OBJECT_ID
                        || s == BUILT_IN_FUNCTION_ID
                        || s == BUILT_IN_JSX_ID
                )
            } else {
                false
            }
        }
        Type::Function { .. } => true,
        _ => false,
    }
}

/// Check if two dependency lists are equal.
fn are_equal_dependencies(
    a: &[ReactiveScopeDependency],
    b: &[ReactiveScopeDependency],
    env: &Environment,
) -> bool {
    if a.len() != b.len() {
        return false;
    }
    for a_val in a {
        let a_decl = env.identifiers[a_val.identifier.0 as usize].declaration_id;
        let found = b.iter().any(|b_val| {
            let b_decl = env.identifiers[b_val.identifier.0 as usize].declaration_id;
            a_decl == b_decl && are_equal_paths(&a_val.path, &b_val.path)
        });
        if !found {
            return false;
        }
    }
    true
}

/// Check if two dependency paths are equal.
fn are_equal_paths(a: &[DependencyPathEntry], b: &[DependencyPathEntry]) -> bool {
    a.len() == b.len()
        && a.iter()
            .zip(b.iter())
            .all(|(ai, bi)| ai.property == bi.property && ai.optional == bi.optional)
}

/// Check if a scope is eligible for merging with subsequent scopes.
fn scope_is_eligible_for_merging(scope_id: ScopeId, env: &Environment) -> bool {
    let scope = &env.scopes[scope_id.0 as usize];
    if scope.dependencies.is_empty() {
        // No dependencies means output never changes — eligible
        return true;
    }
    scope.declarations.iter().any(|(_key, decl)| {
        let ty = &env.types[env.identifiers[decl.identifier.0 as usize].type_.0 as usize];
        is_always_invalidating_type(ty)
    })
}
